Functions/ImageProcessing/xASL_im_MergePWI4D.m

```diff
@@ -79,39 +79,48 @@
-		end
-
-		if isfield(jsonPWI4Dcurrent, 'PostLabelingDelay')
-			if iSession == 1
-				jsonPWI4DConcatenated.PostLabelingDelay = jsonPWI4Dcurrent.PostLabelingDelay;
-			else
-				jsonPWI4DConcatenated.PostLabelingDelay = [jsonPWI4DConcatenated.PostLabelingDelay; jsonPWI4Dcurrent.PostLabelingDelay];
-			end
-		else
-			warning(['Missing PostLabelingDelay for session ' x.modules.asl.sessionsToMerge{iSession}]);
-		end
-
-		if isfield(jsonPWI4Dcurrent, 'LabelingDuration')
-			if iSession == 1
-				jsonPWI4DConcatenated.LabelingDuration = jsonPWI4Dcurrent.LabelingDuration;
-			else
-				jsonPWI4DConcatenated.LabelingDuration = [jsonPWI4DConcatenated.LabelingDuration; jsonPWI4Dcurrent.LabelingDuration];
-			end
-		else
-			jsonPWI4DConcatenated.LabelingDuration = [];
-			if ~strcmpi(x.Q.LabelingType, 'pasl')
-				warning(['Missing LabelingDuration for session ' x.modules.asl.sessionsToMerge{iSession}]);
+		% Take the following list of fields to save to the merged JSON. 
+		% Some fields are concatenated across sessions. Other fields are given per entire sequence and not per volume. So the fields are saved only once for the first session
+		% and with following session, we check for consistence and report errors
+
+		% A list of fields
+		listFields2Merge = {'MagneticFieldStrength', 'PulseSequenceType', 'MRAcquisitionType', 'Manufacturer', 'ArterialSpinLabelingType', ...% These are checked but not concatenated
+			                'LabelingDuration', 'PostLabelingDelay', 'EchoTime'};% These are concatenated
+
+		% An indication of a field is merged or concatenated
+		bFields2Concatenate = [0, 0, 0, 0, 0, ...
+		                       1, 1, 1];
+
+		% Go through all the fields to merge
+		for iField2Merge = 1:length(listFields2Merge)
+			% The field exists and is not empty, we therefore need to check if we can merge it across sessions
+ 			if isfield(jsonPWI4Dcurrent, listFields2Merge{iField2Merge}) && ~isempty(jsonPWI4Dcurrent.(listFields2Merge{iField2Merge}))
+				if iSession == 1 
+					% For the first session, just add the field
+					jsonPWI4DConcatenated.(listFields2Merge{iField2Merge}) = jsonPWI4Dcurrent.(listFields2Merge{iField2Merge});
+				elseif ~isfield(jsonPWI4DConcatenated, listFields2Merge{iField2Merge})
+					% Session > 1, but the field was not added for session 1, we report a conflict
+					error(['Cannot merge sessions as the JSON field ' listFields2Merge{iField2Merge} ' is missing for session ' x.modules.asl.sessionsToMerge{1} ' but not for session ' x.modules.asl.sessionsToMerge{iSession}]);
+				else
+					% The field seems to be consistently present/missing/empty on both sessions. The field is either check for consistency across sessions or concatenated
+					if bFields2Concatenate(iField2Merge)
+						% Concatenation
+						jsonPWI4DConcatenated.(listFields2Merge{iField2Merge}) = [jsonPWI4DConcatenated.(listFields2Merge{iField2Merge}); jsonPWI4Dcurrent.(listFields2Merge{iField2Merge})];
```

Comment:
This is incorrect right?
Session1 = [Volume1 Volume2]

Session2 = [Volume3 Volume4]

Concatenated = [Session1, Session2]

NOT

Concatenated = [Session1; Session2]

Reply:
Correct. That's how things come out of JSON. I've checked that. I had [Session1, Session2] initially, but that wouldn't work.

Reply:
Are you sure? Internally, throughout the ASL module, we use e.g.,

x.module.asl.etc.parameter.PLD = [500 1000 1500 2000];

So if it comes from JSON as 

x.module.asl.etc.parameter.PLD = 
[500
1000
1500
2000];

then it should be transposed right, and we keep everything internally the same (always horizontal, never vertical)

I mean, most important is that we are consistent throughout ExploreASL. And because we always used horizontal rather than vertical for these vectors (and BIDS as well within the JSON file), we should read a JSON to internal memory horizontal. Right?

Reply:
See above. I'm sure. 

Internally, we don't have a convention, we deal with vectors as vectors and always make sure that the code can work with any orientation to avoid issues with vectors are entered by a different orientation by the user.

I can make the code here also work with any orientation if you want. But I wouldn't impose a strict orientation as this is not needed.